### PR TITLE
update metric mAP@k for image retrieval

### PR DESCRIPTION
**Pull request by zzc98**
## Motivation

Support mAP@k in image retrieval

## Use cases 

```python 
    from mmcls.evaluation.metrics import AveragePrecision

    test_num = 2
    k = 5
    pred = torch.zeros((test_num, k))
    target = torch.Tensor([[1, 0, 0, 1, 0],
                           [0, 1, 1, 0, 1]])
    result = AveragePrecision.calculate(pred, target, average='retrieval')
    print(f'mAP@{k} is {result}%.')
    # mAP@5 is 58.47222137451172%.

```

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by Ezra-Yu — APPROVED** [review]
LGTM.
